### ISSUE TITLE
services.xserver doesn't support multiple GPU setups

### ISSUE THREAD

**Issue by LouisTakePILLz** [NixOS/nixpkgs#108018]
## The problem

The current NixOS xserver module generates `Device` and `Screen` sections for each driver listed in `services.xserver.videoDrivers`. While this works fine for detecting which driver to use for single-GPU setups, it's completely incompatible with multi-GPU setups.

e.g:
Input: `services.xserver.videoDrivers = [ "nvidia" "amdgpu" ];`

Resulting X config:
```
Section "ServerLayout"
  Identifier "Layout[all]"

  # Reference the Screen sections for each driver.  This will
  # cause the X server to try each in turn.
  Screen "Screen-nvidia[0]"
  Screen "Screen-amdgpu[0]"
EndSection

# For each supported driver, add a "Device" and "Screen"
# section.

Section "Device"
  Identifier "Device-nvidia[0]"
  Driver "nvidia"
EndSection

Section "Screen"
  Identifier "Screen-nvidia[0]"
  Device "Device-nvidia[0]"
EndSection

Section "Device"
  Identifier "Device-amdgpu[0]"
  Driver "amdgpu"
EndSection

Section "Screen"
  Identifier "Screen-amdgpu[0]"
  Device "Device-amdgpu[0]"
EndSection
```

This also has the unintended side-effect of picking up the first GPU that matches the "Driver" directive (most likely dependent on PCI enumeration order).

While I can't speak for everyone, I personally configure my monitor layout through the Xorg config (other solutions include RandR scripts):
```nix
{
  services.xserver = {
    enable = true;
    videoDrivers = [ "nvidia" ];
    screenSection = ''
      Option "nvidiaXineramaInfoOrder" "DP-4"
      Option "metamodes" "DP-4: nvidia-auto-detect, DP-2: nvidia-auto-detect"
    '';
  };
}
```

However, this doesn't work for multi-GPU configurations because `screenSection` is injected into all `Screen` sections.
Currently, the only way to fix this is to write the entire config by hand, bypassing most of the xserver options:
```nix
{
  services.xserver = {
    enable = true;
    videoDrivers = [ "nvidia" ];
    config = lib.mkForce ''
      # X config here
    '';
  };
}
```

## The solution

In order to address this issue, major refactoring of the xserver module would be required. However, I don't think it's a good idea since a lot of options would get deprecated and/or break backwards compatibility. Perhaps a better idea would be to deprecate `services.xserver` and move the Xorg configuration to a different module? (this ties into #53394, #94799 and [discourse#8051](https://discourse.nixos.org/t/moving-services-xserver-displaymanager-to-services-displaymanager/8051))
The only GPU I have is an NVIDIA one so I can't compare configuration differences. Personally, I don't think it's possible to come up with a good driver-agnostic solution.
Perhaps integrating the monitor configuration into a different namespace would be a better idea? e.g:
```nix
{
  # this is just a random suggestion; there's probably a more appropriate namespace for this
  services.graphics.nvidia-x11 = {
    enable = true;
    devices = {
      gpu0 = {
        gpuId = "GPU-eb1363bc-1951-8e58-1103-03e2a08a3dec.GPU-0";
        busId = "PCI:1:0:0";
        primary = true;
      };
      gpu1 = {
        gpuId = "GPU-70ca8357-d891-40eb-8c90-da3abb2522e0.GPU-0";
        busId = "PCI:2:0:0";
      };
    };
    monitors = [
      {
        device = "gpu0";
        port = "DP-2";
        width = 3440;
        height = 1440;
        offsetX = 1440;
        offsetY = 560;
        refreshRate = 120;
        primary = true;
      }
      {
        device = "gpu1";
        port = "DP-4";
        width = 1440;
        height = 2560;
        refreshRate = 120;
        attributes = {
          Rotation = "right";
        };
      }
    ];
  };
}
```

would generate something like this:
```
# This is not a working example. It's missing some things like the ServerLayout section
Section "Device"
  Identifier "Device-gpu0"
  Driver "nvidia"
  BusID "PCI:1:0:0"
EndSection

Section "Device"
  Identifier "Device-gpu1"
  Driver "nvidia"
  BusID "PCI:2:0:0"
EndSection

Section "Screen"
  Identifier "Screen0"

  # This is decided by which device has `primary = true`
  # (I have no idea what this does when combined with the metamodes directive below)
  Device "Device-gpu0"

  # This is decided by which monitor has `primary = true`
  Option "nvidiaXineramaInfoOrder" "GPU-eb1363bc-1951-8e58-1103-03e2a08a3dec.GPU-0.DP-2"

  # Apparently you can list displays from different GPUs in the same metamodes directive? See https://collaboradev.com/2016/09/30/arch-linux-hydra-build/
  Option "metamodes" "GPU-eb1363bc-1951-8e58-1103-03e2a08a3dec.GPU-0.DP-2: 3440x1440_120 +1440+560, GPU-70ca8357-d891-40eb-8c90-da3abb2522e0.GPU-0.DP-4: 1440x2560_120 +0+0 {Rotation=right}"
EndSection
```

This is all untested as I don't own a second NVIDIA GPU. This topic came up on IRC (@KarlJoad)

Other related issue(s): #96999

**Comment by aanderse** [comment]
Maybe in this scenario the user supplying their own completely custom `xorg.conf` is the best option. I believe that can be achieved with
```
services.xserver.config = lib.mkForce ''
# xorg.conf custom contents
'';
```

**Comment by thiagokokada** [comment]
This is similar with issue https://github.com/NixOS/nixpkgs/issues/75007 (lack of flexibility in multi-devices configuration using `libinput`).

I think a refactoring of the Xserver related modules is really needed, those limitations are kinda bad and surprising for the users.

**Comment by LouisTakePILLz** [comment]
@aanderse even though in this case writing the config manually might be preferable (due to the large amount of nvidia-specific stuff), the xserver module remains very much inflexible.

After seeing how powerful nix can be, it just leaves me wishing the xserver module was just a little better :)

**Comment by aanderse** [comment]
Fair enough. Is this something you're interested in tackling?

**Comment by LouisTakePILLz** [comment]
I'm interested but I'm not sure where to start. There's also the part where the display managers are too tightly coupled with the X server... so I guess the question is more about where does the X server settings belong?

**Comment by symphorien** [comment]
Do you think it would help if there was a `settings` option as in https://github.com/NixOS/rfcs/pull/42 and the existing interface only set settings through the settings options? This way you could use `mkForce` on the relevant sections.

**Comment by aanderse** [comment]
It might take some creativity to coerce `xorg.conf` into nix attributes... but if you could it would be the basis for a great module.

**Comment by thiagokokada** [comment]
> It might take some creativity to coerce xorg.conf into nix attributes... but if you could it would be the basis for a great module.

Yeah, for the most part it doesn't look to hard:

```
Section "Screen"
        Identifier "dummy_screen"
        Device "dummy_card"
        Monitor "dummy_monitor"
        SubSection "Display"
        EndSubSection
EndSection
```

```nix
g
```

But things like Options or Modelines looks like a pain:

```
{
				services.xserver-new.settings = {
								Screen = {
												Identifier = "dummy_screen";
												Device = "dummy_card";
												Monitor = "dummy_monitor";
												Display = {
																SomeOption = "foo";
												};
								};
				};
}
```

**Comment by symphorien** [comment]
it is possible to leave hard things as strings. The goal is to reduce the amount of string that the user with special needs must `mkForce`. 
In you specific case, I'm not xorg conf expert but this looks like reasonable:
```
{
Modeline = { "1920x1080" = [ 172.80 1920 2040 2248 2576 1080 1081 1084 1118 ]; };
Options = { DPI = "96 x 96"; }
}
```

**Comment by thiagokokada** [comment]
@symphorien But in this case we would need to have a special case for Modeline and Options, so they would be blocks right? Not that I am complaining, your idea seems better than mine:

```nix
{
  services.xserver-new.settings = {
    Screen = {
      Identifier = "dummy_screen";
      HorizSync = "28.0-80.0";
      VertRefresh = "48.0-75.0";
      Modeline = [ "1920x1080" 172.80 1920 2040 2248 2576 1080 1081 1084 1118 ];
      Option = [ "DPI" "96 x 96" ];
    };
  };
}
```

**Comment by symphorien** [comment]
Yes, special casing some keys does not sound unreasonable to me. It somewhat increases complexity however, that's the implementer's call to make I guess.
I think `Option = [ "DPI" "96 x 96" ];` is not going to fly if it is expected that you can set option twice (no idea if it's really possible, just gessing)
```
Option "Foo" bar
Option "Baz" bar2
```
does not easily map to an attrset, because you can't set the key "Option" twice in nix. That's the kind of gotchas you need to be careful of when designing RFC 42 `settings` options I think.

**Comment by eadwu** [comment]
You can use arrays for duplicate keys and keep the "array" values as a pure string. So maybe something like
```nix
settings = [
  {  Section = {
     Option = [
       { AllowMouseOpenFail = "on";
          DontZap = "on";
       }
     ];
  }}
];
```

I'm not sure if this is possible, if Xorg parses the configuration in order than it'll be impossible if a `Device` depends on a `Screen` or if any order is to be enforced, it'll probably be a mess of `mkOrder`, `mkBefore`, and `mkAfter` in which case I would just prefer the existing schematic that's less messy.

**Comment by adrian-gierakowski** [comment]
I'd be interested in getting this sorted as I've just added a 2nd gpu (nvidia) to my PC (first one is amd) and I'm not getting video output from the second one. I'm using it mostly for CUDA workloads atm but it would be nice to be able to use it for video at some point, especially that I'm also not able to set `Coolbits` to enable oc and advanced fan control. Anything I could to to help pushing this forward?

**Comment by NorfairKing** [comment]
+1 Would like this too

**Comment by not-matthias** [comment]
Any update on this?
